modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java

```diff
@@ -2687,6 +2694,55 @@ private void _validateGroupId(long groupId, String scope)
 		}
 	}
 
+	private void _validateNextState(
+			long objectFieldId, long listTypeDefinitionId,
+			ObjectEntry objectEntry, Map.Entry<String, Serializable> entry)
+		throws PortalException {
+
+		ListTypeEntry listTypeEntryRequest =
+			_listTypeEntryLocalService.getListTypeEntry(
+				listTypeDefinitionId, String.valueOf(entry.getValue()));
+
+		ObjectStateFlow objectStateFlow =
+			_objectStateFlowLocalService.fetchObjectFieldObjectStateFlow(
+				objectFieldId);
+
+		ObjectState objectStateRequest =
+			_objectStateLocalService.getObjectStateFlowObjectState(
+				listTypeEntryRequest.getListTypeEntryId(),
+				objectStateFlow.getObjectStateFlowId());
+
+		Map<String, Serializable> objectEntryValues = objectEntry.getValues();
+
+		String objectEntryValue = _getValue(
+			String.valueOf(objectEntryValues.get(entry.getKey())));
+
+		ListTypeEntry listTypeEntry =
+			_listTypeEntryLocalService.getListTypeEntry(
+				listTypeDefinitionId, objectEntryValue);
+
+		ObjectState objectState =
+			_objectStateLocalService.getObjectStateFlowObjectState(
+				listTypeEntry.getListTypeEntryId(),
+				objectStateFlow.getObjectStateFlowId());
+
+		List<ObjectState> objectStatesPossibles =
+			_objectStateLocalService.getNextObjectStates(
+				objectState.getObjectStateId());
+
+		Stream<ObjectState> stream = objectStatesPossibles.stream();
+
+		if (!Objects.equals(objectState, objectStateRequest) &&
+			!stream.anyMatch(
+				listPossiblesObjectStates -> Objects.equals(
+					listPossiblesObjectStates.getObjectStateId(),
+					objectStateRequest.getObjectStateId()))) {
+
+			throw new ObjectEntryValuesException.InvalidObjectStateChange(
+				listTypeEntry.getKey(), listTypeEntryRequest.getKey());
+		}
+	}
```

Comment:
@RenanUmV 

I believe we can simplify this method,

we need to get the current value of the picklist state field and get its transition avaiables.

the current value we can get from objectEntry as you did here:

````
Map<String, Serializable> objectEntryValues = objectEntry.getValues();

String objectEntryValue = _getValue(String.valueOf(objectEntryValues.get(entry.getKey())));
````

and the next available transitions of a state we can get by using this method:

[getNextObjectStates](https://github.com/liferay/liferay-portal/blob/master/modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectStateLocalServiceImpl.java#L82)

The second step we just need to check if the `entry.getValue()` is included in next available transitions list, otherwise it should throw an exception.

does it make sense to you?

feel free to contact me if you need help.

Thank you!

Reply:
I could not simplify the method, because it is using the listTypeEntryRequest and objectStateFlow to find the objectStateRequest, then I use the objectEntryValue to find the listTypeEntry, from the listTypeEntry I find the current objectState of the picklist. The method getNextObjectState gets the current objectStateId of the picklist and returns the list of possible Object States that the picklist's objectState can go to, then checks if the objectStateRequest is present in the list returned by getNextObjectStates. The ObjectState table doesn't have a column with the label of the Object State, so I can't compare with `entry.getValue()`.